File: src/commands/discover.ts
import { Command } from 'commander';
import chalk from 'chalk';
import { spawnSync } from 'child_process';
import fs from 'fs';

export const discoverCommand = new Command('discover')
  .description('Run full discovery pipeline: start proxy, generate traffic, diff against spec')
  .requiredOption('-s, --spec <path>', 'path to existing OpenAPI spec to diff against')
  .requiredOption('-t, --target <url>', 'target API URL to proxy to')
  .option('--compose-file <path>', 'docker-compose file for infrastructure', 'docker-compose.yml')
  .option('--traffic-output <path>', 'path to write traffic JSONL', '/tmp/traffic.jsonl')
  .option('--report-output <path>', 'path to write diff report', '/tmp/diff_report.json')
  .option('--token <token>', 'X-Plex-Token for authenticated requests')
  .option('--exit-on-gap', 'exit with code 2 if gaps are found')
  .action(async (options) => {
    // 1. Start docker compose if compose-file exists
    if (fs.existsSync(options.composeFile)) {
      console.log(chalk.blue('Starting docker compose...'));
      const up = spawnSync('docker', ['compose', '-f', options.composeFile, 'up', '-d'], { stdio: 'inherit' });
      if (up.status !== 0) {
        console.error(chalk.red('Failed to start docker compose'));
        process.exit(1);
      }

      // Wait for services to be healthy
      console.log(chalk.blue('Waiting for services...'));
      await new Promise((resolve) => setTimeout(resolve, 15000));
    }

    // 2. Start arbiter proxy in background (we'll use the existing server module)
    // For simplicity, this command assumes the proxy is already running or will be started separately
    console.log(chalk.yellow('Note: This command assumes the Arbiter proxy is running.'));
    console.log(chalk.yellow('Start it in another terminal with: arbiter start -t <target>'));

    // 3. Generate traffic
    console.log(chalk.blue('Generating traffic...'));
    // We'll shell out to ourselves for generate-traffic
    const trafficArgs = ['generate-traffic', '--target', 'http://localhost:8080'];
    if (options.token) {
      trafficArgs.push('--token', options.token);
    }
    const traffic = spawnSync(process.argv[0], [process.argv[1], ...trafficArgs], {
      encoding: 'utf-8',
      stdio: ['inherit', 'pipe', 'inherit'],
    });
    console.log(traffic.stdout);

    // 4. Diff
    console.log(chalk.blue('Running diff...'));
    const diffArgs = ['diff', '--spec', options.spec, '--traffic', options.trafficOutput];
    if (options.reportOutput) {
      diffArgs.push('--output', options.reportOutput);
    }
    if (options.exitOnGap) {
      diffArgs.push('--exit-on-gap');
    }
    const diff = spawnSync(process.argv[0], [process.argv[1], ...diffArgs], {
      encoding: 'utf-8',
      stdio: 'inherit',
    });

    // 5. Stop docker compose
    if (fs.existsSync(options.composeFile)) {
      console.log(chalk.blue('Stopping docker compose...'));
      spawnSync('docker', ['compose', '-f', options.composeFile, 'down'], { stdio: 'inherit' });
    }

    process.exit(diff.status || 0);
  });

File: rust/src/cli/discover.rs
//! `arbiter discover` (port of src/commands/discover.ts).

use std::process::Command;

use clap::Args;

use super::fail;

/// Run full discovery pipeline: start proxy, generate traffic, diff against spec
#[derive(Args)]
pub struct DiscoverArgs {
    /// path to existing OpenAPI spec to diff against
    #[arg(short = 's', long = "spec")]
    pub spec: String,

    /// target API URL to proxy to
    #[arg(short = 't', long = "target")]
    pub target: String,

    /// docker-compose file for infrastructure
    #[arg(long = "compose-file", default_value = "docker-compose.yml")]
    pub compose_file: String,

    /// path to write traffic JSONL
    #[arg(long = "traffic-output", default_value = "/tmp/traffic.jsonl")]
    pub traffic_output: String,

    /// path to write diff report
    #[arg(long = "report-output", default_value = "/tmp/diff_report.json")]
    pub report_output: String,

    /// X-Plex-Token for authenticated requests
    #[arg(long = "token")]
    pub token: Option<String>,

    /// exit with code 2 if gaps are found
    #[arg(long = "exit-on-gap")]
    pub exit_on_gap: bool,
}

pub fn run(args: &DiscoverArgs) -> i32 {
    let compose_active = std::path::Path::new(&args.compose_file).exists();

    if compose_active {
        println!("Starting docker compose...");
        let up = Command::new("docker")
            .args(["compose", "-f", &args.compose_file, "up", "-d"])
            .status()
            .unwrap_or_else(|e| fail(format!("Failed to start docker compose: {e}")));
        if !up.success() {
            fail("Failed to start docker compose");
        }
        println!("Waiting for services...");
        std::thread::sleep(std::time::Duration::from_secs(15));
    }

    println!("Generating traffic...");
    let exe = std::env::current_exe()
        .unwrap_or_else(|e| fail(format!("Failed to locate arbiter binary: {e}")));
    let mut traffic_cmd = Command::new(&exe);
    traffic_cmd
        .args([
            "generate-traffic",
            "--target",
            &args.target,
            "--output",
            &args.traffic_output,
        ])
        .stdin(std::process::Stdio::inherit())
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::inherit());
    if let Some(token) = args.token.as_deref() {
        traffic_cmd.args(["--token", token]);
    }
    let traffic_output = traffic_cmd
        .output()
        .unwrap_or_else(|e| fail(format!("Failed to run generate-traffic: {e}")));
    if !traffic_output.stdout.is_empty() {
        println!("{}", String::from_utf8_lossy(&traffic_output.stdout));
    }

    println!("Running diff...");
    let mut diff_cmd = Command::new(&exe);
    diff_cmd.args([
        "diff",
        "--spec",
        &args.spec,
        "--traffic",
        &args.traffic_output,
    ]);
    if !args.report_output.is_empty() {
        diff_cmd.args(["--output", &args.report_output]);
    }
    if args.exit_on_gap {
        diff_cmd.arg("--exit-on-gap");
    }
    let diff_status = diff_cmd
        .status()
        .unwrap_or_else(|e| fail(format!("Failed to run diff: {e}")));

    if compose_active {
        println!("Stopping docker compose...");
        Command::new("docker")
            .args(["compose", "-f", &args.compose_file, "down"])
            .status()
            .ok();
    }

    diff_status.code().unwrap_or(0)
}
